services/motion/builtin/utils.go

```diff
@@ -0,0 +1,532 @@
+	sens := createMovementSensor(t, ctx, logger, motors, fakeBase, origin, noise)
+
+	deps := resource.Dependencies{}
+	deps[motors[leftMotorName].Name()] = motors[leftMotorName]
+	deps[motors[rightMotorName].Name()] = motors[rightMotorName]
+	deps[fakeBase.Name()] = fakeBase
+	deps[sens.Name()] = sens
+
+	return deps
+}
+
+func createWheeledBase(
+	t *testing.T,
+	ctx context.Context,
+	logger logging.Logger,
+	motors map[string]motor.Motor,
+	geomSize float64,
+) base.Base {
+	t.Helper()
+
+	reg, ok := resource.LookupRegistration(base.API, wheeled.Model)
+	test.That(t, ok, test.ShouldBeTrue)
+
+	wheeledConf := &wheeled.Config{
+		WidthMM:              200,
+		WheelCircumferenceMM: 2000,
+		Left:                 []string{leftMotorName},
+		Right:                []string{rightMotorName},
+	}
+
+	baseConf := resource.Config{
+		Name:                baseName,
+		API:                 base.API,
+		Frame:               &referenceframe.LinkConfig{Geometry: &spatialmath.GeometryConfig{R: geomSize}},
+		ConvertedAttributes: wheeledConf,
+	}
+	deps := resource.Dependencies{}
+	deps[motors[leftMotorName].Name()] = motors[leftMotorName]
+	deps[motors[rightMotorName].Name()] = motors[rightMotorName]
+
+	wBase, err := reg.Constructor(ctx, deps, baseConf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	return wBase.(base.Base)
+}
+
+func createEncoders(t *testing.T, ctx context.Context, logger logging.Logger) map[string]encoder.Encoder {
+	t.Helper()
+	leftconf := resource.Config{
+		Name:                leftMotorName + "_encoder",
+		API:                 encoder.API,
+		ConvertedAttributes: &fakeencoder.Config{UpdateRate: int64(updateRate)},
+	}
+	rightconf := resource.Config{
+		Name:                rightMotorName + "_encoder",
+		API:                 encoder.API,
+		ConvertedAttributes: &fakeencoder.Config{UpdateRate: int64(updateRate)},
+	}
+	left, err := fakeencoder.NewEncoder(ctx, leftconf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	right, err := fakeencoder.NewEncoder(ctx, rightconf, logger)
+	test.That(t, err, test.ShouldBeNil)
+	return map[string]encoder.Encoder{leftMotorName: left, rightMotorName: right}
+}
+
+func createMotors(
+	t *testing.T,
+	ctx context.Context,
+	logger logging.Logger,
+	encoders map[string]encoder.Encoder,
+) map[string]motor.Motor {
+	t.Helper()
+	motorMap := map[string]motor.Motor{}
+	for name, encoder := range encoders {
+		conf := resource.Config{
+			Name:                name,
+			API:                 motor.API,
+			ConvertedAttributes: &fakemotor.Config{Encoder: name + "_encoder", TicksPerRotation: 4000},
+		}
+
+		thisMotor, err := fakemotor.NewMotor(
+			ctx,
+			resource.Dependencies{encoder.Name(): encoder},
+			conf,
+			logger,
+		)
+		test.That(t, err, test.ShouldBeNil)
+		motorMap[name] = thisMotor
+	}
+	return motorMap
+}
+
+type noisyMovementSensor struct {
+	movementsensor.MovementSensor
+	noise      spatialmath.Pose
+	queryCount int // Apply noise every other iteration when this is > 10
```

Comment:
This feels hacky, can we get away without it?

Reply:
So, the annoying part of simulating all of the motion stack on as close to real hardware as we can get, is that it limits our ability to insert arbitrary noise like this.

We want to replan multiple times for some tests. But if we only apply the noise once, then that new noise is now "where the motion service thinks the base is", and it will plan from that point. So unless we have multiple times that noise is applied, there will only be one replan.

I agree this is hacky. I am very open to other ways to do this.

Reply:
On the other hand, this method of doing things is playing havoc with the race condition detector, so I will explore other options